src/eduid_webapp/reset_password/helpers.py

```diff
@@ -0,0 +1,369 @@
+
+    if isinstance(state, ResetPasswordEmailAndPhoneState) and state.phone_code.is_expired(sms_expiration_time):
+        current_app.logger.info(f'Phone code expired for state: {email_code}')
+        # Revert the state to EmailState to allow the user to choose extra security again
+        current_app.password_reset_state_db.remove_state(state)
+        state = ResetPasswordEmailState(eppn=state.eppn, email_address=state.email_address,
+                                        email_code=state.email_code)
+        current_app.password_reset_state_db.save(state)
+        raise BadCode('resetpw.expired-sms-code')
+
+    return state
+
+
+def send_password_reset_mail(email_address: str):
+    """
+    :param email_address: User input for password reset
+    """
+    try:
+        user = current_app.central_userdb.get_user_by_mail(email_address,
+                                                        raise_on_missing=False)
+    except UserHasNotCompletedSignup:
+        # Old bug where incomplete signup users where written to the central db
+        user = None
+    if not user:
+        current_app.logger.info(f"Found no user with the following "
+                                 "address: {email_address}.")
+        return None
+    state = ResetPasswordEmailState(eppn=user.eppn,
+                                    email_address=email_address,
+                                    email_code=get_unique_hash())
+    current_app.password_reset_state_db.save(state)
+    text_template = 'reset_password_email.txt.jinja2'
+    html_template = 'reset_password_email.html.jinja2'
+    to_addresses = [address.email for address in user.mail_addresses.verified.to_list()]
+
+    pwreset_timeout = current_app.config.email_code_timeout // 60 // 60  # seconds to hours
+    context = {
+        'reset_password_link': url_for('reset_password.set_new_pw',
+                                       email_code=state.email_code.code,
+                                       _external=True),
+        'password_reset_timeout': pwreset_timeout
+    }
+    subject = _('Reset password')
+    send_mail(subject, to_addresses, text_template,
+              html_template, current_app, context, state.reference)
+    current_app.logger.info(f'Sent password reset email to user {state.eppn}')
+    current_app.logger.debug(f'Mail addresses: {to_addresses}')
+
+
+def generate_suggested_password() -> str:
+    """
+    The suggested password is hashed and saved in session to avoid form hijacking
+    """
+    password_length = current_app.config.password_length
+
+    password = generate_password(length=password_length)
+    password = ' '.join([password[i*4: i*4+4] for i in range(0, math.ceil(len(password)/4))])
+
+    return password
+
+
+def hash_password(password: str, salt: str,
+                  strip_whitespace: bool = True) -> bytes:
+    """
+    :param password: string, password as plaintext
+    :param salt: string or None, NDNv1H1 salt to be used for pre-hashing
```

Comment:
doc param says string or None, the type hint and code disagrees :)

Reply:
fixed